src/SonarScanner.MSBuild.Shim/AdditionalFilesService.cs

```diff
@@ -45,23 +45,89 @@ public class AdditionalFilesService(IDirectoryWrapper directoryWrapper) : IAddit
+        // Respect user defined sonar.tests and do not re-populate it.
+        // This might lead to some files considered as both source and test, in which case the user should exclude them via sonar.exclusions.
+        return HasUserSpecifiedSonarTests(analysisConfig)
+            ? new(allFiles, [])
+            : PartitionAdditionalFiles(allFiles, analysisConfig);
+    }
 
-        if (extensions.Length > 0)
+    private List<string> GetAllFiles(IEnumerable<string> extensions, DirectoryInfo projectBaseDir)
+    {
+        var pattern = string.Join("|", extensions.Select(x => x.TrimStart('.') + "$").Distinct());
+        var regex = new Regex(pattern, RegexOptions.Compiled | RegexOptions.IgnoreCase, RegexTimeout);
+        return directoryWrapper
+            .EnumerateFiles(projectBaseDir.FullName, "*", SearchOption.AllDirectories)
+            .Where(regex.SafeIsMatch)
+            .ToList();
+    }
+
+    private static bool HasUserSpecifiedSonarTests(AnalysisConfig analysisConfig) =>
+        analysisConfig.LocalSettings.Exists(x => x.Id == "sonar.tests");
+
+    private static AdditionalFiles PartitionAdditionalFiles(List<string> allFiles, AnalysisConfig analysisConfig)
+    {
+        var testExtensions = GetTestExtensions(analysisConfig);
+        if (testExtensions.Length == 0)
         {
-            var pattern = string.Join("|", extensions.Select(x => x.TrimStart('.') + "$").Distinct());
-            var regex = new Regex(pattern, RegexOptions.Compiled | RegexOptions.IgnoreCase, RegexTimeout);
-            return directoryWrapper.EnumerateFiles(projectBaseDir.FullName, "*", SearchOption.AllDirectories).Where(regex.SafeIsMatch);
+            return new(allFiles, []);
         }
-        else
+        var testsPattern = string.Join("|", testExtensions);
+        var testsRegex = new Regex(testsPattern, RegexOptions.Compiled | RegexOptions.IgnoreCase, RegexTimeout);
+        var sources = new List<string>();
+        var tests = new List<string>();
+        foreach (var file in allFiles)
         {
-            return [];
+            if (testsRegex.SafeIsMatch(file))
+            {
+                tests.Add(file);
+            }
+            else
+            {
+                sources.Add(file);
+            }
         }
+        return new(sources, tests);
     }
+
+    private static string[] GetTestExtensions(AnalysisConfig analysisConfig) =>
+        analysisConfig.ServerSettings
+            .Where(x => SupportedTestLanguages.Contains(x.Id))
+            .SelectMany(x => x.Value.Split(','))
+            .SelectMany(x => SupportedTestInfixes.Select(infix => $"{infix}{x}$"))
+            .Distinct()
+            .ToArray();
+
+    private static string[] GetExtensions(AnalysisConfig analysisConfig) =>
+        SupportedLanguages
+            .Select(x => analysisConfig.ServerSettings.Find(property => property.Id == x))
+            .Where(x => x is not null)
+            .SelectMany(x => x.Value.Split(','))
```

Comment:
can the user use a space between the extensions? I'm not sure if we should support or not that scenario